### PR TITLE
Added static methods for classes

### PR DESCRIPTION
**Pull request by tanay-man**

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
The design would the following:
Convert
```py
class test:
    def fn():
        pass
```
to
```c
StructType test  {
    class procedure :: fn => f
    def f:
        pass
}
```

**Comment by Thirumalai-Shaktivel** [comment]
Here is the simple example from LFortran for member function:
```fortran
module m

type t
    integer :: x
    integer :: y
contains
    procedure :: fn => f
end type t

contains

integer function f(this)
class(t), intent(in) :: this
end function

end module

program test
use m
implicit none

type(t) :: t1
print *, t1%fn()

end program test
```

**Comment by Thirumalai-Shaktivel** [comment]
Let's mark it as draft PR until it is ready to be merged.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -2939,77 +2947,102 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
...
+                    abi,
+                    false,
+                    false                    
+                );
+                ASR::symbol_t *cls_proc_sym = ASR::down_cast<ASR::symbol_t>(tmp);
+                current_scope->add_symbol(class_proc_name, cls_proc_sym);
+                continue;
             } else if (AST::is_a<AST::Pass_t>(*x.m_body[i])) {
                 continue;
-            } else if (!AST::is_a<AST::AnnAssign_t>(*x.m_body[i])) {
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:2948]: Remove this changes, it seems very confusing, we will it in another PR

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 2]
```diff
@@ -2939,77 +2947,102 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
...
+                ASR::Function_t* f = ASR::down_cast<ASR::Function_t>(f_sym);
+                std::string class_proc_name = "Xx_Class_Procedure_"+std::string(f->m_name);
+                SymbolTable* proc_scope = ASRUtils::symbol_parent_symtab(f_sym);
+                Str s;
+                s.from_str(al,class_proc_name);
+                ASR::abiType abi = ASR::abiType::Source;
+                tmp = make_ClassProcedure_t(
+                    al,
+                    f_sym->base.loc,
+                    proc_scope,
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:2965]: Here, Keep the ClassProcedure name as f_ast->name and Function name as a class_proc_name,
See my example: https://github.com/lcompilers/lpython/pull/2721#issuecomment-2145533534

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
The implementation seems good, there are minor improvements. We will do it once the MRE works.

Also, for the code generation error, we need to visit StructType in the Module scope in LLVM backend (asr_to_llvm.cpp).
In LFortran the Function is available in the Module scope but here we keep it in the StructType scope, right? So, visit the symbols of StructType in the Module visitor.

**Comment by Thirumalai-Shaktivel** [comment]
```diff
diff --git a/src/lpython/semantics/python_ast_to_asr.cpp b/src/lpython/semantic
s/python_ast_to_asr.cpp
index 01031d704..8c9dd2dee 100644
--- a/src/lpython/semantics/python_ast_to_asr.cpp
+++ b/src/lpython/semantics/python_ast_to_asr.cpp
@@ -3194,12 +3194,14 @@ public:
                                 x.base.base.loc);
         }
         SymbolTable *parent_scope = current_scope;
-        ASR::symbol_t* clss_sym = current_scope->get_symbol(x_m_name);
+        ;
         ASR::StructType_t* clss = nullptr;
-        if(clss_sym != nullptr && !is_enum(x.m_bases, x.n_bases) && !is_union(
x.m_bases, x.n_bases)){
-            clss = ASR::down_cast<ASR::StructType_t>(clss_sym);
-            current_scope = clss->m_symtab;
-            is_generating_body = true;
+        if(ASR::symbol_t* clss_sym = current_scope->get_symbol(x_m_name) ) {
+            if ( is_a<ASR::StructType_t>(*clss_sym) ){
+                clss = ASR::down_cast<ASR::StructType_t>(clss_sym);
+                current_scope = clss->m_symtab;
+                is_generating_body = true;
+            }
```

File: src/libasr/stzY3vVD [hunk 3]
```diff
@@ -0,0 +1 @@
+!<arch>
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/stzY3vVD:1]: Remove this

File: tests/reference/asr-structs_01-0893e35.json [hunk 4]
```diff
@@ -5,9 +5,9 @@
     "infile_hash": "9342dd6345fb71bbbfb1c41e643530c58493758d08774f838b1c85e0",
     "outfile": null,
     "outfile_hash": null,
-    "stdout": null,
-    "stdout_hash": null,
-    "stderr": "asr-structs_01-0893e35.stderr",
-    "stderr_hash": "6f58b337cbc9cb1832e2ecba47fdad4b64e4484b913e29f61c64dbea",
-    "returncode": 2
+    "stdout": "asr-structs_01-0893e35.stdout",
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-structs_01-0893e35.json:8]: This should be an error

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
Add a test and make other tests to pass.

**Comment by Thirumalai-Shaktivel** [comment]
Let's finish off this and move on __init__ in a new PR.

**Comment by tanay-man** [comment]
fixes #2722

File: tests/reference/asr-structs_10-cb8a283.stderr [hunk 5]
```diff
@@ -1,9 +1,5 @@
-semantic error: Struct member functions are not supported
- --> tests/errors/structs_10.py:7:5 - 8:24
+semantic error: The type 'StringIO' is undeclared.
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-structs_10-cb8a283.stderr:1]: I think for struct, we shouldn't allow member functions.

@certik, What do you think?

File: tests/reference/asr-structs_01-0893e35.stdout [hunk 6]
```diff
@@ -0,0 +1,116 @@
+(TranslationUnit
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-structs_01-0893e35.stdout:1]: I think this should be removed.

File: integration_tests/class_01.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from lpython import i32,i64
+
+def fn_2():
+    print("Inside fn_2")
+    return
+ 
+class Test:
+    mem : i64 = i64(5)
+    s   : str = "abc"
+    def fn_1():
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/class_01.py:10]: Print are not being test, so I suggest to pass some arguments to the functions and check the computed return values.

File: integration_tests/class_01.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+         
+def main():
+    t: Test = Test()
+    print(t.mem)
+    assert t.mem == i64(5)
+    print(t.s)
+    assert t.s == "abc"
+    Test.fn_1()
+
+main()
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/class_01.py:28]: Minor: make sure to add a newline at the end of the file.

File: integration_tests/class_01.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+class Test:
+    mem : i64 = i64(5)
+    s   : str = "abc"
+    def fn_1():
+        print("Inside fn_1")
+        print("fn_2 called")
+        fn_2()
+        print("fn_3 called")
+        Test.fn_3()
+        return
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/class_01.py:16]:
```suggestion
        return
```
No extra spaces as well

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 10]
```diff
@@ -3014,6 +3014,30 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+        for (auto &item : x.m_symtab->get_scope()){
+            if (is_a<ASR::Function_t>(*item.second)){
+                ASR::Function_t *v = down_cast<ASR::Function_t>(item.second);
+                instantiate_function(*v);
+            }
+        }
+        current_scope = current_scope_copy;
+    }
+
+    void make_struct_f_def(const ASR::StructType_t &x){
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:3029]: If you have time, try simplifying this function: visit_StructType and make_struct_f_def.
They seem duplicated

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 11]
```diff
@@ -2926,7 +2925,29 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
...
+                    AST::Name_t *n = AST::down_cast<AST::Name_t>(ann_assign->m_target);
+                    std::string var_name = n->m_id;
+                    ASR::expr_t* init_expr = nullptr;
+                    visit_AnnAssignUtil(*ann_assign, var_name, init_expr, false, abi, true);
+                    ASR::symbol_t* var_sym = current_scope->resolve_symbol(var_name);
+                    ASR::call_arg_t c_arg;
+                    c_arg.loc = var_sym->base.loc;
+                    c_arg.m_value = init_expr;
+                    member_init.push_back(al, c_arg);
+                }
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:2947]: This seems repeated as well. 
I will try to look into them

**Comment by certik** [comment]
Static method typically means class method --- it doesn't access any members of the instance. I think in this case you mean a method that is not virtual?

How are class variables implemented in ASR --- can you show how the ASR looks like for the above class?

**Comment by certik** [comment]
Before we go too deep into this, I think we need to use Class for classes, not Struct.

Otherwise it will be hard to undo.

File: src/libasr/asr_verify.cpp [hunk 12]
```diff
@@ -529,11 +529,13 @@ class VerifyVisitor : public BaseWalkVisitor<VerifyVisitor>
                 ASR::is_a<ASR::StructType_t>(*a.second) ||
                 ASR::is_a<ASR::UnionType_t>(*a.second) ||
                 ASR::is_a<ASR::ExternalSymbol_t>(*a.second) ||
-                ASR::is_a<ASR::CustomOperator_t>(*a.second) ) {
+                ASR::is_a<ASR::CustomOperator_t>(*a.second) ||
+                ASR::is_a<ASR::Function_t>(*a.second))  {
```
Comment (certik) [inline src/libasr/asr_verify.cpp:533]: What was the issue here? Why is Function now skipped?

**Comment by Thirumalai-Shaktivel** [comment]
I forgot about the Class symbol and thought to represents the same as LFortran. Sorry about that.
Here is the ASR being generated:

<details><summary>ASR</summary>
<p>

```Clojure
(TranslationUnit
    (SymbolTable
        1
        {
            __main__:
                (Module
                    (SymbolTable
                        2
                        {
                            Test:
                                (StructType
                                    (SymbolTable
                                        4
                                        {
                                            fn_1:
                                                (Function
                                                    (SymbolTable
                                                        5
                                                        {
                                                            Test_fn_3:
                                                                (ExternalSymbol
                                                                    5
                                                                    Test_fn_3
                                                                    4 fn_3
                                                                    Test
                                                                    []
                                                                    fn_3
                                                                    Public
                                                                )
                                                        })
                                                    fn_1
                                                    (FunctionType
                                                        []
                                                        ()
                                                        Source
                                                        Implementation
                                                        ()
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        []
                                                        .false.
                                                    )
                                                    [fn_2]
                                                    []
                                                    [(Print
                                                        [(StringConstant
                                                            "Inside fn_1"
                                                            (Character 1 11 ())
                                                        )]
                                                        ()
                                                        ()
                                                    )
                                                    (Print
                                                        [(StringConstant
                                                            "fn_2 called"
                                                            (Character 1 11 ())
                                                        )]
                                                        ()
                                                        ()
                                                    )
                                                    (SubroutineCall
                                                        2 fn_2
                                                        ()
                                                        []
                                                        ()
                                                    )
                                                    (Print
                                                        [(StringConstant
                                                            "fn_3 called"
                                                            (Character 1 11 ())
                                                        )]
                                                        ()
                                                        ()
                                                    )
                                                    (SubroutineCall
                                                        5 Test_fn_3
                                                        ()
                                                        []
                                                        ()
                                                    )
                                                    (Return)]
                                                    ()
                                                    Public
                                                    .false.
                                                    .false.
                                                    ()
                                                ),
                                            fn_3:
                                                (Function
                                                    (SymbolTable
                                                        6
                                                        {
                                                            
                                                        })
                                                    fn_3
                                                    (FunctionType
                                                        []
                                                        ()
                                                        Source
                                                        Implementation
                                                        ()
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        []
                                                        .false.
                                                    )
                                                    []
                                                    []
                                                    [(Print
                                                        [(StringConstant
                                                            "Inside fn_3"
                                                            (Character 1 11 ())
                                                        )]
                                                        ()
                                                        ()
                                                    )]
                                                    ()
                                                    Public
                                                    .false.
                                                    .false.
                                                    ()
                                                ),
                                            mem:
                                                (Variable
                                                    4
                                                    mem
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Integer 8)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            s:
                                                (Variable
                                                    4
                                                    s
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Character 1 -2 ())
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                )
                                        })
                                    Test
                                    []
                                    [mem
                                    s]
                                    Source
                                    Public
                                    .false.
                                    .false.
                                    [((Cast
                                        (IntegerConstant 5 (Integer 4))
                                        IntegerToInteger
                                        (Integer 8)
                                        (IntegerConstant 5 (Integer 8))
                                    ))
                                    ((StringConstant
                                        "abc"
                                        (Character 1 3 ())
                                    ))]
                                    ()
                                    ()
                                ),
                            __main__global_stmts:
                                (Function
                                    (SymbolTable
                                        8
                                        {
                                            
                                        })
                                    __main__global_stmts
                                    (FunctionType
                                        []
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    [main]
                                    []
                                    [(SubroutineCall
                                        2 main
                                        ()
                                        []
                                        ()
                                    )]
                                    ()
                                    Public
                                    .false.
                                    .false.
                                    ()
                                ),
                            fn_2:
                                (Function
                                    (SymbolTable
                                        3
                                        {
                                            
                                        })
                                    fn_2
                                    (FunctionType
                                        []
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    []
                                    [(Print
                                        [(StringConstant
                                            "Inside fn_2"
                                            (Character 1 11 ())
                                        )]
                                        ()
                                        ()
                                    )
                                    (Return)]
                                    ()
                                    Public
                                    .false.
                                    .false.
                                    ()
                                ),
                            main:
                                (Function
                                    (SymbolTable
                                        7
                                        {
                                            Test_fn_1:
                                                (ExternalSymbol
                                                    7
                                                    Test_fn_1
                                                    4 fn_1
                                                    Test
                                                    []
                                                    fn_1
                                                    Public
                                                ),
                                            t:
                                                (Variable
                                                    7
                                                    t
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Struct
                                                        2 Test
                                                    )
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                )
                                        })
                                    main
                                    (FunctionType
                                        []
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    []
                                    [(Assignment
                                        (Var 7 t)
                                        (StructTypeConstructor
                                            2 Test
                                            [((Cast
                                                (IntegerConstant 5 (Integer 4))
                                                IntegerToInteger
                                                (Integer 8)
                                                (IntegerConstant 5 (Integer 8))
                                            ))
                                            ((StringConstant
                                                "abc"
                                                (Character 1 3 ())
                                            ))]
                                            (Struct
                                                2 Test
                                            )
                                            ()
                                        )
                                        ()
                                    )
                                    (Print
                                        [(StructInstanceMember
                                            (Var 7 t)
                                            4 mem
                                            (Integer 8)
                                            ()
                                        )]
                                        ()
                                        ()
                                    )
                                    (Assert
                                        (IntegerCompare
                                            (StructInstanceMember
                                                (Var 7 t)
                                                4 mem
                                                (Integer 8)
                                                ()
                                            )
                                            Eq
                                            (Cast
                                                (IntegerConstant 5 (Integer 4))
                                                IntegerToInteger
                                                (Integer 8)
                                                (IntegerConstant 5 (Integer 8))
                                            )
                                            (Logical 4)
                                            ()
                                        )
                                        ()
                                    )
                                    (Print
                                        [(StructInstanceMember
                                            (Var 7 t)
                                            4 s
                                            (Character 1 -2 ())
                                            ()
                                        )]
                                        ()
                                        ()
                                    )
                                    (Assert
                                        (StringCompare
                                            (StructInstanceMember
                                                (Var 7 t)
                                                4 s
                                                (Character 1 -2 ())
                                                ()
                                            )
                                            Eq
                                            (StringConstant
                                                "abc"
                                                (Character 1 3 ())
                                            )
                                            (Logical 4)
                                            ()
                                        )
                                        ()
                                    )
                                    (SubroutineCall
                                        7 Test_fn_1
                                        ()
                                        []
                                        ()
                                    )]
                                    ()
                                    Public
                                    .false.
                                    .false.
                                    ()
                                )
                        })
                    __main__
                    []
                    .false.
                    .false.
                ),
            main_program:
                (Program
                    (SymbolTable
                        9
                        {
                            __main__global_stmts:
                                (ExternalSymbol
                                    9
                                    __main__global_stmts
                                    2 __main__global_stmts
                                    __main__
                                    []
                                    __main__global_stmts
                                    Public
                                )
                        })
                    main_program
                    [__main__]
                    [(SubroutineCall
                        9 __main__global_stmts
                        2 __main__global_stmts
                        []
                        ()
                    )]
                )
        })
    []
)
```

</p>
</details>

**Comment by Thirumalai-Shaktivel** [comment]
@tanay-man, as @certik suggested instead of `StructType`, we need to use `ClassType` (later we need to rename it to `Class`).
I think you can work on a new branch.

To check for the `dataclass`, you can use `is_dataclass` function and all the Struct implementation within the if block.

**Comment by tanay-man** [comment]
> Static method typically means class method --- it doesn't access any members of the instance. I think in this case you mean a method that is not virtual?
> 
> How are class variables implemented in ASR --- can you show how the ASR looks like for the above class?

Currently due to 'self' not being implemented, we cannot access instance variables. I think static methods is correct.

**Comment by Thirumalai-Shaktivel** [comment]
Also, @tanay-man, as we have to create ClassType, we need handle everywhere, like in asr_verify, declaration, Constructor, etc.

**Pull request closed without merging**